core/src/main/java/com/oath/oak/ValueUtilsImpl.java

```diff
@@ -32,161 +31,168 @@
+                                     MemoryManager memoryManager, InternalOakMap internalOakMap) {
         int capacity = serializer.calculateSize(newVal);
-        if (capacity + getHeaderSize() > s.getByteBuffer().remaining()) {
-            return moveValue(chunk, lookUp, memoryManager, internalOakMap, newVal);
+        if (capacity > ctx.value.getLength()) {
+            return moveValue(chunk, ctx, memoryManager, internalOakMap, newVal);
         }
-        ByteBuffer bb = getValueByteBufferNoHeader(s);
+        ByteBuffer bb = getValueByteBufferNoHeader(ctx.value);
         serializer.serialize(newVal, bb);
-        return s;
+        return TRUE;
     }
 
-    private <V> Slice moveValue(
-        Chunk<?, V> chunk, EntrySet.LookUp lookUp, MemoryManager memoryManager,
-        InternalOakMap internalOakMap, V newVal) {
+    private <V> ValueResult moveValue(
+            Chunk<?, V> chunk, ThreadContext ctx, MemoryManager memoryManager,
+            InternalOakMap internalOakMap, V newVal) {
 
-        Slice oldSlice = lookUp.valueSlice.duplicate();
-        Slice newSlice = internalOakMap.overwriteExistingValueForMove(lookUp, newVal, chunk);
-        if (newSlice == null) {
+        boolean moved = internalOakMap.overwriteExistingValueForMove(ctx, newVal, chunk);
+        if (!moved) {
             // rebalance was needed or the entry was updated by someone else, need to retry
-            return null;
+            return RETRY;
         }
-        // can not release the old slice or mark it moved, before the new one is updated!
-        setLockState(oldSlice, MOVED);
-        memoryManager.releaseSlice(oldSlice);
-        return newSlice;
+        // can now release the old slice or mark it moved, before the new one is updated!
+        setLockState(ctx.value, MOVED);
+        memoryManager.release(ctx.value);
+
+        ctx.value.copyFrom(ctx.newValue);
+        return TRUE;
     }
 
     @Override
-    public ValueResult compute(Slice s, Consumer<OakWriteBuffer> computer, int version) {
-        ValueResult result = lockWrite(s, version);
+    public ValueResult compute(Slice s, Consumer<OakWriteBuffer> computer) {
+        ValueResult result = lockWrite(s);
         if (result != TRUE) {
             return result;
         }
-        computer.accept(new OakAttachedWriteBuffer(s, getHeaderSize()));
-        unlockWrite(s);
+
+        OakAttachedWriteBuffer writeBuffer = new OakAttachedWriteBuffer(s);
+        try {
+            computer.accept(writeBuffer);
+        } finally {
+            writeBuffer.disable();
+            unlockWrite(s);
+        }
+
         return TRUE;
     }
 
     @Override
-    public <V> Result<V> remove(Slice s, MemoryManager memoryManager, int version, V oldValue,
-                                OakTransformer<V> transformer) {
+    public <V> Result remove(ThreadContext ctx, MemoryManager memoryManager, V oldValue,
```

Comment:
I have encountered a difficulty with this code while debugging. I wrote a debugging structure, for some reason assuming, Result is returned by any ValueUtils interface. Initially I looked on remove, of course. But then I found it is not the case... What can be done about it?

The output result is from different types, thus causing confusion. For example:
Result removeResult = valueOperator.remove(ctx, memoryManager, oldValue, transformer);
Vs
ValueUtils.ValueResult res = valueOperator.compute(ctx.value, computer);

Reply:
It is funny because the class `Result` is a composite of `ValueResult` with an extra value object.

Currently, the logic behind it is that a `Result` object is only returned when we need to return a value on success, in addition to the result status.

We could consider always returning a `Result` object, but then all the `ValueUtils` methods should accept a `Result` or a `ThreadContext` object as an input.
This may result in some inconsistencies: for example, `compute()` might return `TRUE` but the `value` will not be set, which is inconsistent with `remove()` and `transform()`.

What are your thoughts on that?

BTW, do you agree that `TRUE` should be `SUCCESS` and `FALSE` should be `FAIL`?